libr/util/axml.c

```diff
@@ -0,0 +1,281 @@
+
+enum {
+	TYPE_STRING_POOL = 0x0001,
+	TYPE_XML = 0x0003,
+	TYPE_START_ELEMENT = 0x0102,
+	TYPE_END_ELEMENT = 0x103,
+};
+
+enum {
+	RESOURCE_STRING = 0x03,
+};
+
+// Beginning of every header
+R_PACKED (
+typedef struct {
+	ut16 type;
+	ut16 header_size;
+	ut32 size;
+}) chunk_header_t;
+
+// String pool referenced throughout the Binary XML, there must only be ONE
+R_PACKED (
+typedef struct {
+	ut32 string_count;
+	ut32 style_count;
+	ut32 flags;
+	ut32 strings_offset;
+	ut32 styles_offset;
+	ut32 offsets[];
+}) string_pool_t;
+
+R_PACKED (
+typedef struct {
+	ut16 size;
+	ut8 unused;
+	ut8 type;
+	ut32 data;
+}) resource_value_t;
+
+R_PACKED (
+typedef struct {
+	ut32 namespace;
+	ut32 name;
+	ut32 unused;
+	resource_value_t value;
+}) attribute_t;
+
+R_PACKED (
+typedef struct {
+	ut32 line;
+	ut32 comment;
+	ut32 namespace;
+	ut32 name;
+	ut32 flags;
+	ut16 attribute_count;
+	char unused[sizeof(ut16) * 3];
+	attribute_t attributes[];
+}) start_element_t;
+
+R_PACKED (
+typedef struct {
+	ut32 line;
+	ut32 comment;
+	ut32 namespace;
+	ut32 name;
+}) end_element_t;
+
+// TODO: Bounds checking, ensure endianness handling, handle UTF-8
+static char *string_lookup(string_pool_t *pool, const ut8 *data, ut32 i, size_t *length) {
+	if (i > r_read_le32 (&pool->string_count)) {
+		return NULL;
+	}
+
+	ut32 offset = r_read_le32 (&pool->offsets[i]);
+	ut16 *start = (ut16 *)((uintptr_t)data + r_read_le32 (&pool->strings_offset) + 8 + offset);
+
+	// If >0x7fff, stored as a big-endian ut32
+	ut16 n = r_read_le16 (start++);
+	if (n & 0x8000) {
+		n |= ((n & 0x7fff) << 16) | r_read_le16 (start++);
+	}
+
+	// Size of UTF-16LE without NULL
+	n *= 2;
+
+	char *name = calloc(n * 2 + 1, 1);
+
+	// If UTF-16LE, decode to UTF-8 so we can print it to the screen
+	if (r_str_utf16_to_utf8 ((unsigned char*)name, n * 2, (const unsigned char *)start, n, true) < 0) {
+		eprintf ("Failed to decode UTF16-LE\n");
+		return NULL;
+	}
+
+	if (length) {
+		*length = n;
+	}
+
+	return name;
+}
+
+static char *resource_value(string_pool_t *pool, const ut8 *data, resource_value_t *value) {
+	switch (value->type) {
+	case RESOURCE_STRING: {
+		size_t size = 0;
+		char *n = string_lookup (pool, data, r_read_le32 (&value->data), &size);
+		// Space for quotes and NULL
+		char *f = calloc (size + 2 + 1, 1);
+		snprintf (f, size + 2 + 1, "\"%s\"", n);
+		free (n);
+		return f;
+	} break;
+	default:
+		// TODO: Handle all resource value types
+		break;
+	}
+	return NULL;
+}
+
+ut32 depth = 0;
+
+static bool dump_element(RStrBuf *sb, string_pool_t *pool, const ut8 *data,
+		start_element_t *element, bool start) {
+	char *name = string_lookup (pool, data, r_read_le32(&element->name), NULL);
+	for (ut32 i = 0; i < depth; i++) {
+		r_strbuf_appendf (sb, "\t");
+	}
+
+	if (start) {
+		r_strbuf_appendf (sb, "<%s", name);
+		ut16 count = r_read_le16 (&element->attribute_count);
+		if (count != 0) {
+			r_strbuf_appendf (sb, " ");
+		}
+		for (ut16 i = 0; i < count; i++) {
```

Comment:
Move the ut16 i outside the for statement, thats a c99 feature we prefer to avoid for portability, some old compilers dont handle it.